### PR TITLE
Implement OAuth2 password flow

### LINKED ISSUE #807: Oauth2 password flow
Hi. I'm trying to use the oauth2 feature with password flow from swagger 2.0 spec. I dont' know if I'm doing thing wrong or if there some missing stuff, but let me explain. This is my spec:

```
swagger: '2.0'

info:
  title: BakPak API
  description: BakPak server API
  version: '1.0.0'

host: localhost:5000
basePath: /api/v1

schemes:
  - http
  - https

consumes:
  - application/json

produces:
  - application/json

securityDefinitions:
  basicPassword:
    type: oauth2
    flow: password
    tokenUrl: http://localhost:5000/api/v1/token
    scopes:
      s1: Scope 1
      s2: Scope 2

paths:
  /token:
    get:
      tags:
        - auth
      summary: Get an authentication token
      description: Get an authentication token
      operationId: GetToken
      responses:
        '200':
          description: ""
        default:
          description: Operation failed.
  /test:
    get:
      summary: Test endpoint
      description: Test endpoint
      responses:
        '200':
          description: ""
        default:
          description: Operation failed.
      security:
        - basicPassword:
          - s1
          - s2  
  /user:
    post:
      tags:
        - user
      summary: Add a user
      description: Add a user
      operationId: AddUser
      parameters:
        - name: body
          in: body
          description: Type can be `bp`, `facebook` or `google`.
          schema:
            $ref: '#/definitions/Credentials'
      responses:
        '200':
          description: ""
        default:
          description: Operation failed.

definitions:
  Credentials:
    properties:
      type:
        type: string
      username:
        type: string
      password:
        type: string
      token:
        type: string
```

I have a `/test` endpoint just to test oauth2. I remove the comment in `index.html` to enable oauth2 flow:

```
            initOAuth({
              clientId: "your-client-id",
              realm: "your-realms",
              appName: "your-app-name"
            });
```

When I click on the **"on/off"** button to authorize, a new windows browser is opened with the following url `http://localhost:5000/swagger-ui/null&redirect_uri=http://localhost:5000/swagger-ui/o2c.html&realm=your-realms&client_id=your-client-id&scope=s1,s2`.

Note the `null` in the url. It is normal? I'm not super familiar with every aspect of oauth, but I was expecting a popup for username/password and somehow receiving at some endpoint a `POST` with a 
`application/x-www-form-urlencoded` in the request body like this:
`grant_type=password&username=johndoe&password=A3ddj3w`

reference: http://tools.ietf.org/html/rfc6749#section-4.3.2

### PR DESCRIPTION
**Pull request by MugeSo**
This PR is an OAuth2 password flow implementation works with #2014 

This PR updates #1853 and #1574
And this closes #807

### CODE REVIEW

**Comment by webron** [comment]
@bodnia can you review?

**Comment by hkulekci** [comment]
@MugeSo at least, client_id is necessary for password grant_type request at OAuth2. Right now, Your update is not sending any client information.

**Comment by hkulekci** [comment]
https://github.com/MugeSo/swagger-ui/pull/1 I sent you an pr.

**Comment by MugeSo** [comment]
@hkulekci I cannot find what you say though I read RFC6749 again.
I found only that `clinent_id` is required for client password authentication(for client authentication), authorization code grant or implicit grant.
And password grant does not requires client authentication. it just optional.

**Comment by MugeSo** [comment]
And there is no spec about oauth2 client authentication in Open API specification.
So I send issue https://github.com/OAI/OpenAPI-Specification/issues/785.

**Comment by hkulekci** [comment]
@MugeSo you are right. RFC don't force. But I don't understand how the api understand the people come which client in this scenario.

To be able to work this scenario, we must take a client token before sending a request for a password credential request.

**Comment by MugeSo** [comment]
@hkulekci Ok, I see what you want.
I think we need to implement [client authentication](https://tools.ietf.org/html/rfc6749#section-2.3) correctly, to solve your problem.

And it also solves #2183.

@webron @fehguy Which do you prefers implement client authentication in this PR or as another PR?

**Comment by HugoMario** [comment]
I edited the PR (still not check in) in order to test this flow with the sample oauth2 provider that it's being developed. Adding next headers:

`var requestHeaders = {
            'Authorization': 'Basic ' + btoa(username + ':' + password),
            'Content-Type': 'application/x-www-form-urlencoded'
        };
`

`$.ajax({
            ...
            headers: requestHeaders,
         ...`

Helped me to make flow works and tests oauth2 provider properly.

**Comment by wimpers** [comment]
It seems that the milestone to which https://github.com/openvstorage/framework/issues/1021 belongs is closed. Does this mean that the issue is fixed (I assume not as it is still open) or will this PR be merged in a  newer version?

**Comment by vpandey-rms** [comment]
I too am eagerly watching this PR, this is going to fix so many workarounds we had to implement.

**Comment by wimpers** [comment]
@fehguy , @bodnia  I know you guys must be super busy but could it be that this one wasn't merged in v2.1.4 (https://github.com/swagger-api/swagger-ui/issues/807) ? Can it be merged into the latest version?

**Comment by bundabrg** [comment]
Could not client_id and client_secret be optional parameters asked on the same dialogue as the username/password? That's how I'm implementing it and is far closer to how a real client will work. Saves having to hard-code it in.

**Comment by frol** [comment]
@fehguy @bodnia @webron Can we here at least a word from you? There were already two PRs (#1574, #1853) which got outdated just because of the project refactoring. I don't want to see this one become outdated once again. I beg you to merge it while it has no conflicts.

**Comment by bodnia** [comment]
@wimpers @bundabrg @frol I'll look into it by Monday EOD, sorry for delays.

**Comment by frol** [comment]
@MugeSo @hkulekci The RFC doesn't say a word about `client_id` and `client_secret` for Resource Owner Password Credentials Grant, but the example in section (4.3.2) indicates the `Authorization: Basic czZCaGRSa3F0MzpnWDFmQmF0M2JW` header, which is the `base64(cliend_id:client_secret)` just like @HugoMario implemented.

**Comment by bundabrg** [comment]
Actually I'm pretty confused here. The RFC doesn't state anything about client_id or client_secret for a Resource Owner Password Credentials Grant. Yet nearly every other page I've researched about OAuth2 grants described this grant with client_id and client_secret as part of the exchange, including oauthlib requiring it.

**Comment by frol** [comment]
Actually, the section 3.2.1 states the following:

> **Confidential clients or other clients issued client credentials MUST** authenticate with the authorization server as described in Section 2.3 when making requests to the token endpoint.

It seems to me that the reason why `client_id` is specifically mentioned in some of the flows is that the headers cannot be easily set for those requests (e.g. if you want to perform OAuth2 in pure HTML (without JS) you will only need to define a form with a "hidden" `client_id` field).

**Comment by MugeSo** [comment]
I implement client authentication for password and application flow, because some libraries requires it as you say.

**Comment by frol** [comment]
I have just tested the latest changes and both "Basic auth" and "Request body" work for me on my Flask-Oauthlib based OAuth2 server!

@MugeSo Great job!

**Comment by frol** [comment]
@MugeSo Is there and should be there a way to set a default `client_id`?

**Comment by frol** [comment]
@MugeSo It used to be it, but since that code seems to be dead and mostly useless, I wouldn't mind if the way to configure the defaults gets changed.

**Comment by frol** [comment]
Hmm, I am not sure what is the origin of the problems, so it might be out of the scope of this issue, but let me put them here:

1. The checkbox selection of the scopes is ignored, clicking the "Authorize" button always results in sending out all available scopes.
2. When I added `<div id="auth_container"></div>` container to have a top-level "Authorize" button (like on the top bar here: http://petstore.swagger.io/), the listed scopes include a "vendorExtensions - [object Object]" scope at the end of the scopes list. (I don't experience this issue on the demo site passing my `swagger.json` there)

@MugeSo Do you have any guesses?

**Comment by wimpers** [comment]
@bodnia any update on the issue? It seems more people have chipped and got it working?

**Comment by frol** [comment]
@danballance I think you may try and test it against your server (I have tested it against Flask-OAuthlib and everything works for me as far as I can tell, but it is always better to have an extra pair of eyes here, I guess).

@MugeSo @bodnia A conflict in `AuthView.js` appeared. :(

**Comment by frol** [comment]
By the way, there is a demo API server which bundles this patch rebased on top of a recent master branch: https://github.com/frol/flask-restplus-server-example. The demo is written in Python, but it is really easy to evaluate it if you use Docker:

```sh
$ docker run -it --rm --publish 5000:5000 frolvlad/flask-restplus-server-example
```

http://127.0.0.1:5000/api/v1/ (login: `root`, password: `q`)

**Comment by MugeSo** [comment]
@frol  Thank you for mention. I resolved it :)

File: src/main/javascript/view/Oauth2Model.js [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 SwaggerUi.Models.Oauth2Model = Backbone.Model.extend({
     defaults: {
-        scopes: {}
+        scopes: {},
+        isPasswordFlow: false,
+        clientAuthenticationType: 'none'
```
Comment (frol) [inline src/main/javascript/view/Oauth2Model.js]: Can this default be configurable?

File: src/main/javascript/view/Oauth2Model.js [hunk 2]
```diff
@@ -19,6 +21,13 @@ SwaggerUi.Models.Oauth2Model = Backbone.Model.extend({
             attributes.scopes = scopes;
             this.attributes = attributes;
         }
+
+        if (this.attributes && this.attributes.flow) {
+            var flow = this.attributes.flow;
+            this.set('isPasswordFlow', flow === 'password');
+            this.set('requireClientAuthentication', flow === 'application');
+            this.set('clientAuthentication', flow === 'password' || flow === 'application');
```
Comment (frol) [inline src/main/javascript/view/Oauth2Model.js]: It seems to me that `clientAuthentication` should be available in every OAuth2 Flow.

Comment (MugeSo) [inline src/main/javascript/view/Oauth2Model.js]: Yes, it's advisable what you say.
But, this PR does not support client authentication for other flows.
I add client authentication just to make password flow usable.
I just want to merge password-flow implementation quickly!!

**Comment by wimpers** [comment]
@bodnia any update on the issue? It seems more people have chipped and got it working?

**Comment by danballance** [comment]
Any joy with this folks? Is this functionality available in master now or am I getting ahead of myself?

**Comment by vpandey-rms** [comment]
Please merge this guys, I am waiting for this fix for a long time :)

**Comment by ncarlier** [comment]
+1

**Comment by bodnia** [comment]
@webron I tested, and this works ok and can be merged

**Comment by webron** [comment]
Thanks for all the work and tweaks, @MugeSo - glad to see such a contribution to the project.

**Comment by frol** [comment]
WOW! Supercalifragilisticexpialidocious!

**Comment by hkulekci** [comment]
woow. Finally merged. Thank you.

**Comment by VenomAV** [comment]
I do not know if this is the right place for this question: is there any way to specify default client_id and client_secret for password flow through Swagger UI configurations?

**Comment by bodnia** [comment]
@VenomAV one of options is from https://github.com/swagger-api/swagger-ui/issues/3010 and one more option is with configs, but not implemented yet

**Pull request merged**